src/dbnode/storage/index/convert/convert.go

```diff
@@ -173,6 +173,55 @@ func FromMetricIterNoClone(id ident.ID, tags ident.TagIterator) (doc.Document, e
 	}, nil
 }
 
+// TagsFromTagsIter returns an ident.Tags from a TagIterator. It also tries
+// to re-use bytes from the seriesID if they're also present in the tags
+// instead of re-allocating them. This requires that the ident.Tags that is
+// returned will have the same (or shorter) life time as the seriesID,
+// otherwise the operation is unsafe.
+func TagsFromTagsIter(
+	seriesID ident.ID,
+	iter ident.TagIterator,
+	idPool ident.Pool,
+) (ident.Tags, error) {
+	var (
+		seriesIDBytes = ident.BytesID(seriesID.Bytes())
+		tags          = idPool.Tags()
+	)
+
+	for iter.Next() {
+		curr := iter.Current()
+
+		var (
+			nameBytes, valueBytes = curr.Name.Bytes(), curr.Value.Bytes()
+			tag                   ident.Tag
+			idRef                 bool
+		)
+		if idx := bytes.Index(seriesIDBytes, nameBytes); idx != -1 {
+			tag.Name = seriesIDBytes[idx : idx+len(nameBytes)]
+			idRef = true
+		} else {
+			tag.Name = idPool.Clone(curr.Name)
+		}
+		if idx := bytes.Index(seriesIDBytes, valueBytes); idx != -1 {
+			tag.Value = seriesIDBytes[idx : idx+len(valueBytes)]
+			idRef = true
+		} else {
+			tag.Value = idPool.Clone(curr.Value)
+		}
+
+		if idRef {
+			tag.NoFinalize() // Taken ref, cannot finalize this
+		}
+
+		tags.Append(tag)
+	}
+
+	if err := iter.Err(); err != nil {
+		return ident.Tags{}, err
+	}
+	return tags, nil
```

Comment:
Can we close the iter too before finishing?  When we call this method we usually `.Duplicate()` and we need to close the duplicate.